aten/src/ATen/native/cpu/AdaptiveMaxPoolKernel.cpp

```diff
@@ -0,0 +1,341 @@
+
+  // parallel on dim of N, C
+  at::parallel_for(0, channels, 0, [&](int64_t begin, int64_t end) {
+    for (int64_t c = begin; c < end; c++) {
+      scalar_t* input_ptr = input_data + c * input_height * input_width;
+      scalar_t* output_ptr = output_data + c * output_height * output_width;
+      int64_t* indices_ptr = indices_data + c * output_height * output_width;
+
+      for (int64_t oh = 0; oh < output_height; oh++) {
+        int64_t ih0 = start_index(oh, output_height, input_height);
+        int64_t ih1 = end_index(oh, output_height, input_height);
+
+        for (int64_t ow = 0; ow < output_width; ow++) {
+          int64_t iw0 = start_index(ow, output_width, input_width);
+          int64_t iw1 = end_index(ow, output_width, input_width);
+
+          // compute local max
+          int64_t maxindex = ih0 * input_width + iw0;
+          scalar_t maxval = -std::numeric_limits<scalar_t>::infinity();
+          for (int64_t ih = ih0; ih < ih1; ih ++) {
+            for (int64_t iw = iw0; iw < iw1; iw ++) {
+              int64_t index = ih * input_width + iw;
+              scalar_t val = input_ptr[index];
+              if ((val > maxval) || std::isnan(val)) {
+                maxval = val;
+                maxindex = index;
+              }
+            }
+          }
+
+          // set output to local max and store location of max
+          output_ptr[oh * output_width + ow] = maxval;
+          indices_ptr[oh * output_width + ow] = maxindex;
+        }
+      }
+    }
+  });
+
+  if (!output_.is_contiguous()) {
+    output_.copy_(output);
+  }
+  if (!indices_.is_contiguous()) {
+    indices_.copy_(indices);
+  }
+}
+
+template <typename scalar_t>
+void cpu_adaptive_max_pool_channels_last(
+    Tensor& output_,
+    Tensor& indices_,
+    const Tensor& input_,
+    IntArrayRef output_size) {
+  TORCH_CHECK(input_.ndimension() == 4,
+              "adapative max pooling with channels last format supports tensors with 4 dims");
+  auto memory_format = at::MemoryFormat::ChannelsLast;
+  auto input = input_.contiguous(memory_format);
+  auto output = output_.contiguous(memory_format);
+  auto indices = indices_.contiguous(memory_format);
+
+  auto input_data = input.data_ptr<scalar_t>();
+  auto output_data = output.data_ptr<scalar_t>();
+  auto indices_data = indices.data_ptr<int64_t>();
+
+  int64_t nbatch = input.size(0);
+  int64_t channels = input.size(1);
+  int64_t input_height = input.size(2);
+  int64_t input_width = input.size(3);
+  int64_t output_height = output_size[0];
+  int64_t output_width = output_size[1];
+
+  using Vec = vec256::Vec256<scalar_t>;
+  using integer_t = vec256::int_same_size_t<scalar_t>;
+  using iVec = vec256::Vec256<integer_t>;
+  // for the convience of vectorization, use integer of the same size of scalar_t,
+  //   e.g. int32_t for float, int64_t for double
+  // need to make sure doesn't overflow
+  TORCH_CHECK(input_height * input_width < std::numeric_limits<integer_t>::max());
```

Comment:
I think this check won't work properly if input_height * input_width is bigger than int64_t max since the multiply result might be a small number which will still pass this test. I am not sure whether we will really have such big numbers in reality though, I think probably use `input_height <= std::ceiling(max() / (double)input_width)` is safer.

Reply:
@glaringlee Updated!
Indeed the check is mostly unnecessary since normally input_height * input_width won't be that big. Add this check only because I need to downcast indexing from int64_t to int32_t (for vectorization purpose).